File: JAVA/Pertemuan 04/src/id/ac/uph/pbo/pertemuan04/RekeningTabungan.java
package id.ac.uph.pbo.pertemuan04;

public class RekeningTabungan {
    // private: tidak bisa diakses langsung dari luar
    private String nomorRekening;
    private String namaPemilik;
    private double saldo;
    private int    jumlahTransaksi;

    // Konstruktor
    RekeningTabungan(String nomorRekening, String namaPemilik, double saldoAwal) {
        this.nomorRekening   = nomorRekening;
        this.namaPemilik     = namaPemilik;
        this.jumlahTransaksi = 0;
        this.setSaldo(saldoAwal);   // validasi lewat setter
    }

    // Getter (akses baca)
    String getNomorRekening() { return nomorRekening; }
    String getNamaPemilik()   { return namaPemilik; }
    double getSaldo()         { return saldo; }
    int    getJumlahTransaksi() { return jumlahTransaksi; }

    // Setter dengan validasi
    private void setSaldo(double saldo) {
        if (saldo >= 0) this.saldo = saldo;
        else System.out.println("  ⚠ Saldo awal tidak boleh negatif!");
    }

    // Business method (bukan sekedar getter/setter)
    void setor(double jumlah) {
        if (jumlah > 0) {
            saldo += jumlah;
            jumlahTransaksi++;
            System.out.printf("  [SETOR]  Rp%,12.0f → Saldo: Rp%,12.0f%n", jumlah, saldo);
        } else {
            System.out.println("  ⚠ Jumlah setor harus > 0");
        }
    }

    void tarik(double jumlah) {
        if (jumlah <= 0) {
            System.out.println("  ⚠ Jumlah tarik harus > 0");
        } else if (jumlah > saldo) {
            System.out.printf("  ⚠ Saldo tidak cukup! Saldo: Rp%,.0f%n", saldo);
        } else {
            saldo -= jumlah;
            jumlahTransaksi++;
            System.out.printf("  [TARIK]  Rp%,12.0f → Saldo: Rp%,12.0f%n", jumlah, saldo);
        }
    }

    void cetakInfo() {
        System.out.println("  ┌── Info Rekening ──────────────────────────┐");
        System.out.println("  │ No. Rekening : " + nomorRekening);
        System.out.println("  │ Pemilik      : " + namaPemilik);
        System.out.printf ("  │ Saldo        : Rp%,.0f%n", saldo);
        System.out.println("  │ Transaksi    : " + jumlahTransaksi + " kali");
        System.out.println("  └───────────────────────────────────────────┘");
    }
}

File: PYTHON/Pertemuan 04/RekeningTabungan.py
import sys
sys.stdout.reconfigure(encoding='utf-8')
class RekeningTabungan:
    def __init__(self, nomor_rekening: str, nama_pemilik: str, saldo_awal: float):
        self._nomor_rekening = nomor_rekening
        self._nama_pemilik = nama_pemilik
        self._jumlah_transaksi = 0
        self._saldo = 0.0
        self.saldo = saldo_awal   # validasi lewat property setter

    # Getter (akses baca)
    @property
    def nomor_rekening(self) -> str:
        return self._nomor_rekening

    @property
    def nama_pemilik(self) -> str:
        return self._nama_pemilik

    @property
    def saldo(self) -> float:
        return self._saldo

    @property
    def jumlah_transaksi(self) -> int:
        return self._jumlah_transaksi

    # Setter dengan validasi
    @saldo.setter
    def saldo(self, saldo: float):
        if saldo >= 0:
            self._saldo = saldo
        else:
            print("  ⚠ Saldo awal tidak boleh negatif!")

    # Business method (bukan sekedar getter/setter)
    def setor(self, jumlah: float):
        if jumlah > 0:
            self._saldo += jumlah
            self._jumlah_transaksi += 1
            print(f"  [SETOR]  Rp{jumlah:12,.0f} → Saldo: Rp{self._saldo:12,.0f}")
        else:
            print("  ⚠ Jumlah setor harus > 0")

    def tarik(self, jumlah: float):
        if jumlah <= 0:
            print("  ⚠ Jumlah tarik harus > 0")
        elif jumlah > self._saldo:
            print(f"  ⚠ Saldo tidak cukup! Saldo: Rp{self._saldo:,.0f}")
        else:
            self._saldo -= jumlah
            self._jumlah_transaksi += 1
            print(f"  [TARIK]  Rp{jumlah:12,.0f} → Saldo: Rp{self._saldo:12,.0f}")

    def cetak_info(self):
        print("  ┌── Info Rekening ──────────────────────────┐")
        print(f"  │ No. Rekening : {self._nomor_rekening}")
        print(f"  │ Pemilik      : {self._nama_pemilik}")
        print(f"  │ Saldo        : Rp{self._saldo:,.0f}")
        print(f"  │ Transaksi    : {self._jumlah_transaksi} kali")
        print("  └───────────────────────────────────────────┘")

